### ISSUE TITLE
Capabilities in unrolled contexts

### ISSUE THREAD

**Issue by rachitnigam** [cucapra/seashell#50]
An array access needs to be able to generate sufficient resources to be consumed in an unrolled context:
```
decl a: bit<10>[3 bank 3][3 bank 3];
for (i ...) unroll 3 {
  for (j ...) unroll 3 {
     a[i][j] // correct
     ---
     a[0][j]
     ---
     a[j][i] //correct
  }
}
```
A non-unrolled context implicitly unrolls by `1` which means constants are allowed since they generate exactly one copy of the access:
```
for(i ...) {
  a[0] // correct
  for (j ...) unroll 3 {
    a[0][j] //correct I think?
    a[j][0] // correct I think?
    a[i][j] // correct I think
    a[i][0] // wrong
  }
}
```

**Comment by rachitnigam** [comment]
@sampsyo does the second example make sense?

**Comment by sampsyo** [comment]
Yo—I believe all of this makes sense (and I agree with the correct/incorrect comments on the accesses in the second code listing).

I'm not quite sure what you mean by _generating_ resources, however—it seems like each of these has to decide which resources to _consume_ (but I bet that's what you meant). So in general, we need an (abstract) function that takes a vector of types of index values in an indexing expression and produces a set of banks to consume. This function may also need to take the "iterator stack" as an input?? Nailing this function down "on paper" could be useful.

**Comment by sa2257** [comment]
Hi,

Why is the access `a[I][0]` wrong? Isn't it accessing a unique bank?

And is the function you refer to, the function that determines which banks to access with unrolling? I don't quite follow why the iterator stack is needed?

**Comment by sampsyo** [comment]
Note that the inner loop in that example is unrolled, but the outer loop is not. So `a[j][0]` accesses distinct banks in each of the three parallel "copies" of the inner loop body, but `a[i][0]` accesses the same location three times in parallel. Does that help?

**Comment by sa2257** [comment]
Yes, but are we referring to write accesses? Because accessing the same bank, same location for read is valid.

**Comment by sampsyo** [comment]
Yeah, you can imagine these being writes if you like (where this is definitely a problem). But in the way we are defining our capability-based type system, this is actually a problem for both reads and writes: to get the "special property" of reads you're referring to, where you can access the same location multiple times, we still need to decide when it's safe to _acquire the read capability_ (which would happen in the outer loop here). Fortunately, acquiring capabilities works the same way for both reads and writes; it's just that read capabilities can then be reused multiple times (unlike write capabilities).

**Comment by sa2257** [comment]
But why should acquiring read capability happen in the outer loop? `i` and `0` would guarantee it's accessing a single element regardless of the `unroll`?

**Comment by sampsyo** [comment]
To rewind a little, the idea behind the capability-based design was to split the process of checking accesses into two parts:

- Acquiring the capability. This is meant to be the same for both reads and writes: we can use one, unifying rule to check whether the indexing expression is safe. Acquiring a capability (regardless of type) consumes the appropriate banks so the same capability can't be acquired again.
- Using the capability. Now that you have the capability, you can read or write through it—and those are handled differently. A read capability can be used multiple times, and a write capability can only be used once. This step no longer has to worry about indexing expressions; it just has to worry about the difference between reads and writes.

The goal here is simplicity. The first step has a simple job (consuming the appropriate banks), and the second step has a simple job (permitting the right "number" of accesses to a given location. When they were conflated (before we invented the capability idea), we had one check that kind of had to do _both_ jobs, which made the reasoning more complex.

I should write this stuff up in more detail… I know we've talked about it in person a lot, but I don't think we've written down a full explanation yet (unless I'm forgetting something).

**Comment by sa2257** [comment]
Yes, but should't the error in multiple writes happen at _using the capability_? Doesn't that happen inside the inner loop? Ie. for default `unroll 1` both cases will pass because accessing once is valid inside the outer loop. And writes will fail for `unroll 3` because it was used more than allowed inside the inner loop?

**Comment by sampsyo** [comment]
Got it; I see what you're saying! I guess another way of putting it is that `a[j][0]` can (and should) be promoted outside of the inner loop, while `a[i][0]` cannot. So if we require capability acquisitions to go as early as possible, then we would indeed detect this error at capability _use_ time rather than capability _acquisition_ time.

That could indeed be simpler! Maybe we can get away with _not_ using the iterator-stack information to do this check, then, if we just prohibit unrolled uses of write capabilities that are acquired outside the loop (which correspond to multiple uses of the write capability, which are already disallowed).

**Comment by rachitnigam** [comment]
Operationally, let's think of the capability system and this system separately:
- This system makes sure that an access in an unrolled context uses idx expressions that can provide enough resources. A resource is "provided" by an index type. If the index type has the static range `s`, then it gives `s` resources. For `a[i][j][k]`, you get `s_i * s_j * s_k` resources.
- The capability system manages reads and writes _after_ this check is completed.

**Comment by sa2257** [comment]
@sampsyo yes, I think so. But wouldn't that prohibition also prohibit writes if inner loop is not unrolled? Ie. default `unroll 1`?

@rachitnigam Do you mean this is a separate issue? But why a resource constraint fail `a[I][0]`? It requires 1 resource and `I` provides 1 resource?

**Comment by sampsyo** [comment]
To revise my answer to the above:

> Should't the error in multiple writes happen at _using the capability_?

Yes, that's certainly true—we need to prevent that. But we _also_ (before we even use the capability) need to decide when it's valid to _acquire_ capabilities. So if you like, interpret all of the above examples, wherever there's an indexing expression `a[m][n]`, as saying "acquire a capability for `a[m][n]`." You could imagine that the program does this in preparation for yet another nested loop (not shown in the examples) that will actually use the capability.

**Comment by sa2257** [comment]
An attempt to write down our discussion. Correct me if I’m wrong @rachitnigam .

In the opening example, `a[i][0]` is wrong because the type rule which enforces array access in unrolled loops and type rule for capabilities become complicated. So a work around would be to access `a` in the relevant scope of index access and transfer the value over a local variable.

However, in the example we considered, matrix-matrix multiply (or simplification to matrix vector multiply), this work around no longer works if we try to unroll in multiple contexts. An attempt to use `combine` showed these issues are orthogonal and is also not a work around.

**Comment by rachitnigam** [comment]
So, TLDR, can we write a matrix-vector multiply with two unrolls?

**Comment by rachitnigam** [comment]
Mostly fixed. Subtle issue described in #58.
